core/provider/docker/provider.go

```diff
@@ -0,0 +1,358 @@
+	NetworkCIDR  string   `json:"network_cidr"`
+	AllocatedIPs []string `json:"allocated_ips"`
+
+	BuilderImageName string `json:"builder_image_name"`
+}
+
+type Provider struct {
+	state   *ProviderState
+	stateMu sync.Mutex
+
+	dockerClient           *client.Client
+	dockerNetworkAllocator *ipallocator.Range
+	networkMu              sync.Mutex
+	logger                 *zap.Logger
+}
+
+var _ provider.ProviderI = (*Provider)(nil)
+
+func CreateProvider(ctx context.Context, logger *zap.Logger, providerName string) (*Provider, error) {
+	dockerClient, err := client.NewClientWithOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = dockerClient.Ping(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if providerName == "" {
+		return nil, fmt.Errorf("provider name cannot be empty")
+	}
+
+	state := ProviderState{
+		Name:             providerName,
+		BuilderImageName: "busybox:latest",
+		NetworkName:      fmt.Sprintf("petri-network-%s", providerName),
+		TaskStates:       make(map[string]*TaskState),
+	}
+
+	dockerProvider := &Provider{
+		dockerClient: dockerClient,
+		state:        &state,
+		logger:       logger,
+	}
+
+	network, err := dockerProvider.initNetwork(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerProvider.state.NetworkID = network.ID
+
+	_, cidrMask, err := net.ParseCIDR(network.IPAM.Config[0].Subnet)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerProvider.state.NetworkCIDR = cidrMask.String()
+
+	dockerProvider.dockerNetworkAllocator, err = ipallocator.NewCIDRRange(cidrMask)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return dockerProvider, nil
+}
+
+func RestoreProvider(ctx context.Context, state []byte) (*Provider, error) {
+	var providerState ProviderState
+
+	err := json.Unmarshal(state, &providerState)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dockerProvider := &Provider{
+		state: &providerState,
+	}
+
+	dockerClient, err := client.NewClientWithOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = dockerClient.Ping(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dockerProvider.dockerClient = dockerClient
+	_, cidrMask, err := net.ParseCIDR(providerState.NetworkCIDR)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cidr mask from state: %w", err)
+	}
+
+	dockerProvider.dockerNetworkAllocator, err = ipallocator.NewCIDRRange(cidrMask)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ip allocator from state: %w", err)
+	}
+
+	for _, ip := range providerState.AllocatedIPs {
+		if err := dockerProvider.dockerNetworkAllocator.Allocate(net.ParseIP(ip)); err != nil {
+			return nil, fmt.Errorf("failed to restore ip allocator state: %w", err)
+		}
+	}
+
+	if err := dockerProvider.ensureNetwork(ctx); err != nil {
+		return nil, fmt.Errorf("failed to reconcilliate the docker network: %w", err)
+	}
+
+	return dockerProvider, nil
+}
+
+func (p *Provider) CreateTask(ctx context.Context, definition provider.TaskDefinition) (provider.TaskI, error) {
```

Comment:
It's probably best to return the concrete type `Task` from this function.  Downstream users can always treat it like the `TaskI` interface, but other users can optionally use the concrete type instead of needing to cast to it

Reply:
hmm, actually not necessarily - since this will make the struct not conform to the interface